cpp/util/etcd_delete.cc

```diff
@@ -0,0 +1,127 @@
+#include "util/etcd_delete.h"
+
+#include <functional>
+#include <gflags/gflags.h>
+#include <glog/logging.h>
+#include <mutex>
+
+using std::move;
+using std::mutex;
+using std::pair;
+using std::placeholders::_1;
+using std::string;
+using std::unique_lock;
+using std::vector;
+using util::Status;
+using util::Task;
+using util::TaskHold;
+
+DEFINE_int32(etcd_delete_concurrency, 4,
+             "number of etcd keys to delete at a time");
+
+namespace cert_trans {
+
+namespace {
+
+
+class DeleteState {
+ public:
+  DeleteState(EtcdClient* client, vector<pair<string, int64_t> >&& keys,
+              Task* task)
+      : client_(CHECK_NOTNULL(client)),
+        task_(CHECK_NOTNULL(task)),
+        outstanding_(0),
+        next_(0),
+        keys_(move(keys)) {
+    CHECK_GT(FLAGS_etcd_delete_concurrency, 0);
+
+    if (keys_.empty()) {
+      // Nothing to do!
+      task_->Return();
+    } else {
+      StartNextRequest(unique_lock<mutex>(mutex_));
+    }
+  }
+
+  ~DeleteState() {
+    CHECK_EQ(outstanding_, 0);
+  }
+
+ private:
+  void RequestDone(Task* child_task);
+  void StartNextRequest(unique_lock<mutex>&& lock);
+
+  EtcdClient* const client_;
+  Task* const task_;
+  mutex mutex_;
+  int outstanding_;
+  int next_;
+  const vector<pair<string, int64_t> > keys_;
+};
+
+
+void DeleteState::RequestDone(Task* child_task) {
+  // If a child task has an error, return that error, and do not start
+  // any more requests.
+  if (!child_task->status().ok()) {
+    task_->Return(child_task->status());
+    return;
+  }
+
+  unique_lock<mutex> lock(mutex_);
+  --outstanding_;
+
+  if (next_ < keys_.size()) {
+    StartNextRequest(move(lock));
+  } else {
+    if (outstanding_ < 1) {
+      // No more keys to get, and this was the last one to complete.
+      lock.unlock();
+      task_->Return();
+    }
+  }
+}
+
+
+void DeleteState::StartNextRequest(unique_lock<mutex>&& lock) {
+  CHECK(lock.owns_lock());
+
+  if (task_->CancelRequested()) {
+    // In case the task uses an inline executor.
+    lock.unlock();
+    task_->Return(Status::CANCELLED);
+    return;
+  }
+
+  while (outstanding_ < FLAGS_etcd_delete_concurrency &&
+         next_ < keys_.size()) {
+    CHECK(lock.owns_lock());
+    const pair<string, int64_t>& key(keys_[next_]);
```

Comment:
You could make `next_` a `vector<pair<string, int64_t>>::const_iterator` and then not need this line. (more for clarity if you think it'd help, I realise this is O(1) anyway.)